backend/src/relay.ts

```diff
@@ -224,6 +204,30 @@ export default class Relay {
     }
   }
 
+  private async getBlockWithJustification(rawBlock: SignedBlockJsonRpc) {
+    // adding justifications from finality proof if there is none
+    if (!rawBlock.justifications) {
+      const { number } = rawBlock.block.header;
+      const proof = await pRetry(
+        () => this.sourceApi.rpc.grandpa.proveFinality(number),
+        createRetryOptions(error => this.logger.error(error, `get block justifications for #${number} retry error:`)),
+      );
+      const proofBytes = proof.toU8a(true);
+      // finality proof returns encoded block hash, justification and unknown headers
+      // only care about justification, removing leading extra bytes: 32 for block hash and 4 for SCALE prefix(?)
+      const justification = proofBytes.slice(36);
+
+      rawBlock.justifications = [
+        [
+          [70, 82, 78, 75], // FRNK
```

Comment:
What is `FRNK`?

Reply:
It's engine ID, there are two of them in consensus logs: BABE and FRNK. I don't know the story behind the name, but this is the one we need for GRANDPA verification

Reply:
Named constant like this would be more readable:
```typescript
const GRANDPA_ENGINE_ID = [70, 82, 78, 75]; // FRNK
```
See https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/primitives/finality-grandpa/src/lib.rs#L57-L58